src/propagators.jl

```diff
@@ -0,0 +1,204 @@
+using QuBase
+using Docile
+
+import Base: start, next, done
+
+abstract QuPropagatorMethod
+
+immutable QuPropagator{QPM<:QuPropagatorMethod}
+    hamiltonian
+    init_state::QuVector
+    tlist
+    method::QPM
+    QuPropagator(hamiltonian, init_state, tlist, method) = new(hamiltonian, init_state, tlist, method)
+end
+
+QuPropagator{QPM<:QuPropagatorMethod}(hamiltonian, init_state, tlist, method::QPM) = QuPropagator{QPM}(hamiltonian, init_state, tlist, method)
+
+immutable QuPropagatorState
+    psi::QuVector
+    t
+    t_state
+end
+
+@doc """
+Input Parameters : QuPropagator
+
+Returns the starting iterator state of the propagator method, i.e., the initial state of the system
+""" ->
+function Base.start(prob::QuPropagator)
+    t_state = start(prob.tlist)
+    t,t_state = next(prob.tlist,t_state)
+    return QuPropagatorState(prob.init_state,t,t_state)
+end
+
+@doc """
+Input Parameters : QuPropagator and QuPropagator State
+
+Returns the next state by dispatching to particular
+`propagate` method depending on the `numerical` method type.
+""" ->
+function Base.next{QPM<:QuPropagatorMethod}(prob::QuPropagator{QPM}, qustate::QuPropagatorState)
+    op = prob.hamiltonian
+    current_qustate = qustate.psi
+    current_t = qustate.t
+    t,t_state = next(prob.tlist, qustate.t_state)
+    dt = t - current_t
+    next_qustate = propagate(prob.method, op, dt, current_qustate)
+    return (t, next_qustate), QuPropagatorState(next_qustate, t, t_state)
+end
+
+@doc """
+Input Parameters : QuPropagator and QuPropagator State
+
+Returns true if the current state is final state, else false
+""" ->
+Base.done(prob::QuPropagator, qustate::QuPropagatorState) = done(prob.tlist, qustate.t_state)
+
+@doc """
+Euler Method
+
+Step Propagation using the Euler formula.
+""" ->
+immutable QuEuler <: QuPropagatorMethod
+end
+
+@doc """
+Crank Nicolson Method
+
+Step Propagation using the Crank Nicolson formula.
+""" ->
+immutable QuCN <: QuPropagatorMethod
+end
+
+@doc """
+Krylov subspace Method
+Input Parameters :
+`options` : Dictionary to set the basis size with keyword NC.
+
+For ex :
+x = [:NC => basis_size]
+QuKrylov(x)
+Step Propagation using the Krylov subspace iterations.
+""" ->
+immutable QuKrylov <: QuPropagatorMethod
+    options::Dict
+end
+
+@doc """
+Propagates to the next time state
+Input Parameters:
+`prob`             :  Method to be used
+`op`               :  Hamiltonian of the system
+`t`                :  Time corresponding to the t_state
+`current_t`        :  Current time
+`current_qustate`  :  Quantum state corresponding to current time
+""" ->
+function propagate(prob::QuEuler, op, dt, current_qustate)
+    return (eye(op)-im*op*dt)*current_qustate
+end
+
+function propagate(prob::QuCN, op, dt, current_qustate)
+    uni = eye(op)-im*op*dt/2
+    return \(uni', uni*current_qustate)
+end
+
+function propagate(prob::QuKrylov, op, dt, current_qustate)
+    basis_size = get(prob.options,:NC, length(current_qustate))
```

Comment:
I guess `getkey` gives the key rather than the value ... I had done this so :
```julia
julia> dc = [:NC=>10]
Dict{Symbol,Int64} with 1 entry:
  :NC => 10

julia> getkey(dc, :NC, 0)
:NC

julia> get(dc, :NC, 0)
10
```